src/os/unix/ngx_pipe.c

```diff
@@ -493,5 +525,306 @@ ngx_open_pipe(ngx_cycle_t *cycle, ngx_open_pipe_t *op)
+    ngx_cpystrn((u_char *) ngx_os_argv[0], op->cmd, title_len);
+#endif
+
+    for (;;)
+    {
+        if (ngx_terminate == 1) {
+            return;
+        }
+
+        n_bytes_read = ngx_read_fd(0, read_buf, read_buf_len);
+        if (n_bytes_read == 0) {
+            return;
+        }
+        if (errno == EINTR) {
+            continue;
+        } else if (n_bytes_read < 0) {
+            return;
+        }
+
+        ngx_time_update();
+
+        if (log_fd >= 0) {
+            if (rbcf.interval > 0) {
+                gettimeofday(&tv, &tz);
+                tv.tv_sec -= tz.tz_minuteswest * 60;
+                rbcf.time_now = tv.tv_sec;
+                if ((rbcf.time_now / rbcf.interval) >
+                        (rbcf.last_open_time / rbcf.interval)) {
+                    //need check rollback
+                    ngx_close_file(log_fd);
+                    log_fd = NGX_INVALID_FILE;
+                    ngx_log_error(NGX_LOG_INFO, cycle->log, 0,
+                                  "need check rollback time [%s]", rbcf.logname);
+                    ngx_pipe_do_rollback(cycle, &rbcf);
+                }
+            }
+        }
+
+        if (log_fd >= 0 && rbcf.log_max_size > 0 &&
+                           rbcf.log_size >= rbcf.log_max_size) {
+            ngx_close_file(log_fd);
+            log_fd = NGX_INVALID_FILE;
+            ngx_log_error(NGX_LOG_INFO, cycle->log, 0,
+                          "need check rollback size [%s] [%d]",
+                          rbcf.logname, rbcf.log_size);
+            ngx_pipe_do_rollback(cycle, &rbcf);
+        }
+
+        if (log_fd < 0) {
+            ngx_pipe_create_subdirs(rbcf.logname, cycle);
+            log_fd = ngx_open_file(rbcf.logname, NGX_FILE_APPEND, NGX_FILE_CREATE_OR_OPEN,
+                          NGX_PIPE_FILE_ACCESS);
+            if (log_fd < 0) {
+                ngx_log_error(NGX_LOG_EMERG, cycle->log, ngx_errno,
+                              "open [%s] failed", rbcf.logname);
+                return;
+            }
+
+            gettimeofday(&tv, &tz);
+            tv.tv_sec -= tz.tz_minuteswest * 60;
+            rbcf.last_open_time = tv.tv_sec;
+            if (0 == ngx_fd_info(log_fd, &sb)) {
+                rbcf.log_size = sb.st_size;
+            }
+        }
+
+        if (ngx_write_fd(log_fd, read_buf, n_bytes_read) != n_bytes_read) {
+            if (errno != EINTR) {
+                ngx_log_error(NGX_LOG_EMERG, cycle->log, ngx_errno,
+                              "write to [%s] failed", rbcf.logname);
+                return;
+            }
+        }
+        rbcf.log_size += n_bytes_read;
+    }
+
+}
+
+ngx_int_t
+ngx_pipe_rollback_parse_args(ngx_cycle_t *cycle, ngx_open_pipe_t *op,
+    ngx_pipe_rollback_conf_t *rbcf)
+{
+    u_char         **argv;
+    ngx_uint_t       i;
+    ngx_int_t        j;
+    size_t           len;
+    ngx_str_t        filename;
+    ngx_str_t        value;
+
+    if (op->argv->nelts < 2) {
```

Comment:
The `if (op->argv->nelts < 2) `  It's should be `if (op->argv->nelts < 3) `.
Because `ngx_conf_open_pipe` will push a `NULL` item at the last.
BTW in this scene `ngx_strlen(filename.data)`  will core dump.